parlai/tasks/taskmaster/agents.py

```diff
@@ -0,0 +1,290 @@
+                convo_cheat_sheet[tm_utils.LAST_USER],
+            )
+        else:
+            # ASSISTANT then USER mode [Second pass]
+            episode_idx -= len(
+                self.episode_map["U"]
+            )  # Didn't use '%' because the two maybe unequal in length
+            true_idx = self.episode_map["A"][episode_idx]
+            conversation = self.convos[true_idx]
+            convo_cheat_sheet = self.ep_cheat_sheet[true_idx]
+            first_entry, last_entry = (
+                convo_cheat_sheet[tm_utils.FIRST_ASSISTANT],
+                convo_cheat_sheet[tm_utils.LAST_ASSISTANT],
+            )
+
+        starts_at_odd = first_entry % 2 != 0
+        if starts_at_odd:
+            predecessor = conversation[entry_idx * 2 + 1]['text']
+            successor = conversation[entry_idx * 2 + 2]['text']
+            ep_done = entry_idx * 2 + 1 == last_entry
+        else:
+            predecessor = conversation[entry_idx * 2]['text']
+            successor = conversation[entry_idx * 2 + 1]['text']
+            ep_done = entry_idx * 2 == last_entry
+
+        action = {
+            'id': self.id,
+            'text': predecessor,
+            'episode_done': ep_done,
+            'labels': [successor],
+        }
+
+        return action
+
+
+class SelfDialogueSegmentTeacher(FixedDialogTeacher):
+    """
+    Teacher for written two-person dialogues with labels being relevant/useful
+    parts in the input sentence. The different datatypes of the labels within
+    the data have also been encoded as `label_types`
+    """
+
+    def __init__(self, opt, shared=None):
+        super().__init__(opt)
+        opt['fn'] = "self-dialogs.json"
+
+        if 'exclude-invalid-data' not in opt:
+            opt['exclude-invalid-data'] = True
+
+        if shared and 'convos' in shared:
+            # another instance was set up already, just reference its data
+            self.convos = shared['convos']
+        else:
+            # need to set up data from scratch
+            data_path = tm_utils._path(opt)
+            self._setup_data(data_path, opt)
+
+        self.reset()
+
+    def get(self, episode_idx, entry_idx):
+        conversation = self.convos[episode_idx]
+        conv_len = len(conversation['utterances'])
+        utterance = conversation['utterances'][entry_idx]['text']
+
+        # Check if episode is complete
+        ep_done = entry_idx == conv_len - 1
+        action = {'id': self.id, 'text': utterance, 'episode_done': ep_done}
+
+        # Setup Labels as "text" from segments
+        action['labels'] = []
+        action['label_types'] = []
+        segments = conversation['utterances'][entry_idx]["segments"]
+        for segment in segments:
+            action['labels'] += [segment["text"]]
+            tmp = []
+            for annot in segment["annotations"]:
+                tmp += [annot["name"]]
+            action['label_types'] += [tmp]
+
+        return action
+
+    def num_examples(self):
```

Comment:
ideally you could cache this in an attribute, e.g. `self.num_exs`, which you can initialize as `None` in `__init__` and then return here if not `None` (otherwise set according to logic you have here). this avoids recomputing this number every time `num_examples()` is called (which is a lot)

Reply:
Will do that